apps/server/lib/index.ts

```diff
@@ -42,35 +52,112 @@ process.on('unhandledRejection', (error) => {
-				cluster.fork().on('exit', startWorker);
-			};
-			startWorker();
+		for (let i = 0; i < maxWorkers; i++) {
+			cluster.fork();
 		}
-
 		cluster.on('exit', (worker, code, signal) => {
-			logger.warn(DEFAULT_CONTEXT, `worker ${worker.process.pid} died`, {
-				code,
-				signal,
-			});
+			if (worker.exitedAfterDisconnect === true) {
+				logger.info(
+					context,
+					`worker ${worker?.process?.pid} exited (${signal || code}).`,
+				);
+			} else {
+				logger.info(
+					context,
+					`worker ${worker?.process?.pid} died (${
+						signal || code
+					}). Forking again`,
+				);
+				cluster.fork();
+			}
 		});
+
+		cluster.on('online', (worker) => {
+			logger.debug(
+				context,
+				`Worker ${worker.id} responded after it was forked`,
+			);
+		});
+
+		logger.info(
+			context,
+			`PID: ${process.pid}. Waiting for ${workersNeeded} workers to start`,
+		);
+
+		cluster.on('message', (worker, message) => {
+			if (message?.msg === 'worker-started') {
+				activeWorkers++;
+				logger.info(
+					context,
+					`Worker ${worker.id} worker-started. activeWorkers ${activeWorkers}`,
+				);
+				if (activeWorkers === workersNeeded) {
+					logger.info(context, `All ${workersNeeded} needed workers started.`);
+					handoverPeer.startBroadcasting();
+				}
+			} else if (message?.msg === 'DONE') {
+				// Ignored, is handled by the shutdown code
+			} else {
+				logger.warn(context, `Unknown message received from worker`, message);
+			}
+		});
+
+		// handoverPeer will call this function when we're shutting down. It will `await` until it returns to signal that the handover is done.
+		// Reminder that the handover ( new instance takes the hostname ) is not performed until the old container is killed, so we have to balance between clean-shutdown and almost-zero-downtime
+		// Here we go with the "almost-zero-downtime": we forward the message to all workers and return without waiting for their response.
+		const shutdownCallback = async () => {
+			let exitedWorkers = 0;
+			const liveWorkers = Object.values(cluster.workers || {});
+			for (const worker of liveWorkers) {
+				worker?.on('message', (message) => {
+					if (message?.msg === 'DONE') {
+						exitedWorkers++;
+					}
+				});
+				worker?.send('SHUTDOWN');
+			}
+			// Keep loggins while we're alive
+			setImmediate(async () => {
+				while (exitedWorkers < liveWorkers.length) {
+					logger.info(
+						context,
+						`exitedWorkers: ${exitedWorkers} of ${liveWorkers.length}`,
+					);
+					await new Promise((r) => setTimeout(r, 100));
+				}
+				logger.info(
+					context,
+					`All workers exited. ExitedWorkers: ${exitedWorkers} of ${liveWorkers.length}`,
+				);
+			});
+		};
+
+		handoverPeer.startListening(shutdownCallback);
 	} else {
-		const id = uuidv4();
 		const context = {
-			id: `SERVER-${packageJSON.version}-${environment.pod.name}-worker#${cluster.worker?.id}-${id}`,
+			id: `SERVER-ID-${'[' + serverId + ']'}-worker#${cluster.worker?.id}`,
```

Comment:
We should try to keep the `packageJSON.version` in the logger context.

Reply:
fixed, keeping the packageJSON and adding the new serverId